numpy/core/src/multiarray/datetime.c

```diff
@@ -1761,6 +1762,9 @@ parse_datetime_unit_from_string(char *str, Py_ssize_t len, char *metastr)
                 return NPY_FR_as;
         }
     }
+    else if (allow_generic && len == 7 && !strncmp(str, "generic", 7)) {
```

Comment:
So this works and is simpler
```
    else if (len == 7 && strncmp(str, "generic", 7) == 0) {
        return NPY_FR_GENERIC;
    }
```
Note that you can now do
```
In [4]: np.dtype('M8[generic]')
Out[4]: dtype('<M8')
```
So the question is, why did you have `allow_generic`? Note that the error messages are also incorrect for substrings as they expect null terminated strings.